src/pybind/mgr/poolsets/module.py

```diff
@@ -0,0 +1,1600 @@
+                    # limit so we're good to go.
+                    pass
+                else:
+                    # We can successfully creat the pool within
+                    # target_pgs_per_osd, great
+                    pass
+
+        # Generate pool names
+        for pool_intent in pool_intents:
+            if pool_intent.suffix:
+                # We use "." to join names, because it happens to
+                # conveniently match the RGW convention for zone pools
+                pool_intent.name = self._unique_pool_name(
+                    "{0}.{1}".format(pool_set_name, pool_intent.suffix))
+            else:
+                pool_intent.name = self._unique_pool_name(pool_set_name)
+
+        self._creating = True
+
+        # Finally, we've created all our pools, let's create the
+        # corresponding PoolSet and save it.
+        pool_set = PoolSet()
+        pool_set._name = pool_set_name
+        pool_set.policy = PoolSet.POLICY_AUTOSCALE
+        pool_set._application = {application: {}}
+
+        # TODO: persist these inside poolset and have a recovery
+        # path that applies any intents when we reload, in case
+        # of errors
+        pool_set.intents = pool_intents
+
+        self._poolsets[pool_set._name] = pool_set
+
+        self.log.debug("Saving new poolset")
+        self._mark_dirty()
+        self._save()
+
+        self._do_create(pool_set_name, pool_intents)
+
+        # do_create populates pool IDs
+        for pi in pool_intents:
+            pp = PoolProperties()
+            pp.target_size = pi.target_size
+            pp.target_ratio = pi.target_ratio
+            assert pi.pool_id is not None
+            pool_set.pool_properties[pi.pool_id] = pp
+        pool_set.intents = []
+
+        self.log.debug("{0} {1}".format(
+            len(pool_set.pool_properties),
+            len(pool_set.intents)
+        ))
+
+        self.log.debug("Re-saving with Pool IDs")
+        self._mark_dirty()
+        self._save()
+
+        # TODO: recovery path, if we persist the poolset but
+        # do not finish the actual pool creations
+
+        self._creating = False
+
+        # If we have a conflict with existing consumptions, now is when
+        # we adjust the other poolsets to resolve it.
+        # Classic example is: someone has created a workload designed to
+        # take 100% of the cluster, and now we are coming to create
+        # another workload that we want to share the cluster 50:50
+        # with the existing one.
+        # So: we need to shrink the other guy, set *our* target size to
+        # 50% of the cluster, and then create our pools with baby pg_num,
+        # wait until the other guys are done shrinking, and then grow
+        # our pg_num to the correct target.
+
+        # TODO: track pool creation as an adjustment, and block
+        # any other adjustments until the pool creation is done.
+
+        return 0, '', 'Created poolset {0}'.format(pool_set_name)
```

Comment:
The wip-volumes code relies on this module and it says:

```
            # FIXME: make poolsets return the names, instead of using           
            # magic knowledge of the pattern                                    
            mdp_name = ps_name + ".meta"                                        
            dp_name = ps_name + ".data"                                         
```

It seems like we ought to have this function returning something more meaningful?